### PR TITLE
[MRG] Doing bounds checking on `--num` via command line

### LINKED ISSUE #1243: do bounds checking on --scaled and --num via command line
e.g. `--scaled 0` works => produces a num sketch, which is technically ok but...

### PR DESCRIPTION
**Pull request by keyabarve**
Fixes #1243

Done:
- Add checks for the values of num. num values should always be > 0. if they are unreasonable, say > 50,000 or < 50, we should emit a warning.
- Tests for `watch` command written in tests/test_sourmash.py`.
- Tests for `compute` command written in `tests/test_sourmash_compute.py`.
- Tests for `downsample` command written in `tests/test_cmd_signature.py`.

NOTE:
- `sourmash compute` has been changed to `sourmash sketch` in almost all of the places in `tests/test_sourmash.py` (check #1419).
- The tests for each command have been written in the same file as other tests for that particular command.

### CODE REVIEW

**Comment by keyabarve** [comment]
@ctb I have written the function in `src/sourmash/cli/utils.py` similar to the function I had written for `--scaled`. I have used the `subparser.add_argument` code block in `src/sourmash/cli/sig/downsample.py` as reference and changed the code in this file accordingly. I'm a little confused about what needs to be done with the code in `src/sourmash/cli/compute.py` and `src/sourmash/cli/watch.py` because this is different from the code in `src/sourmash/cli/sig/downsample.py` in the sense that this involved `--num-hashes` instead of `--num`. I was thinking of adding another `parser.add_argument(..)` code block in the function that I wrote, taking these as references.

**Comment by ctb** [comment]
You could do a couple of things - you could add some parameters to the function to set the argument name, or you could add aliases so that `--num` and `--num-hashes` both work, or you could use the `check_num_bounds` function directly in the various files rather than having the `add_num_arg` function.

**Comment by keyabarve** [comment]
What do you mean by: "you could add aliases so that `--num` and `--num-hashes` both work"?

**Comment by keyabarve** [comment]
@ctb What does the `sketch_args.add_argument(...)` in `src/sourmash/cli/compute.py` mean? How is it different from `subparser.add_argument(...)`? Will I have to write my code differently for this file specifically / are there any special things that need to be considered while replacing the code block in this file?

**Comment by keyabarve** [comment]
@ctb I added a few things to the `add_num_arg` function and all the tests now pass. Please review.

**Comment by ctb** [comment]
if it's ready for review and merge, please rename it to MRG, thx

**Comment by ctb** [comment]
> @ctb What does the `sketch_args.add_argument(...)` in `src/sourmash/cli/compute.py` mean? How is it different from `subparser.add_argument(...)`? Will I have to write my code differently for this file specifically / are there any special things that need to be considered while replacing the code block in this file?

I don't know offhand and don't have time to dig into it. If you're unsure and don't want to look into it yourself, you can just leave it alone.

**Comment by ctb** [comment]
> What do you mean by: "you could add aliases so that `--num` and `--num-hashes` both work"?

see https://stackoverflow.com/questions/50896992/argparse-add-argument-alias - the idea is that you could support both styles of argument for each command.

So I could imagine adding `--num` as an alias for `--num-hashes` where the command supports the latter already, and `--num-hashes` as an alias for `--num` where the command supports `--num` already. (You have to do this in order to keep the same CLI behavior working, which we do within major versions of sourmash;  it's fine to add new behavior tho, as long as the old behavior keeps working.)

In practice, if a command supports `--num-hashes`, it will support `--num` because (as I only recently learned myself :) argparse will do substring matching. But I don't think it hurts to explicitly add `--num` as an alias to commands that support `--num-hashing` already.

**Comment by keyabarve** [comment]
> > @ctb What does the `sketch_args.add_argument(...)` in `src/sourmash/cli/compute.py` mean? How is it different from `subparser.add_argument(...)`? Will I have to write my code differently for this file specifically / are there any special things that need to be considered while replacing the code block in this file?
> 
> I don't know offhand and don't have time to dig into it. If you're unsure and don't want to look into it yourself, you can just leave it alone.
>
I'll look into it for a while, but if I'm not being able to find anything useful, I'll leave it alone.

**Comment by keyabarve** [comment]
@ctb I haven't written any tests to test the function I wrote. I'll write some tests in `tests/test_sourmash.py`, using the tests I wrote for `--scaled` as a reference and then put it up for review and merge!

**Comment by keyabarve** [comment]
In this issue, I added the function in `utils.py` and modified `compute.py`, `downsample.py` and `watch.py`.
- Do I need to change any other files?
- When I write tests to check the functionality, which commands should I be testing for?
- Since I changed `sourmash compute` to `sourmash sketch` in almost all the places in the other issue I was working on, do I still need to test the `--num bounds` for `compute`?
- I’ll surely write some tests for the `watch` command. I’m also not sure how to test for the changes in `downsample.py`.
- Should I also test for other commands such as `gather`, `search`?

**Comment by ctb** [comment]
> In this issue, I added the function in `utils.py` and modified `compute.py`, `downsample.py` and `watch.py`.
> 
>     * Do I need to change any other files?

I don't know; do your current changes address the issue?

>     * When I write tests to check the functionality, which commands should I be testing for?

You should cover all of the code that you changed with tests, as well as providing tests to make sure that the original issue was resolved.

>     * Since I changed `sourmash compute` to `sourmash sketch` in almost all the places in the other issue I was working on, do I still need to test the `--num bounds` for `compute`?

That's a good question. Please note it in the top of the PR and it can be part of the review.

>     * I’ll surely write some tests for the `watch` command. I’m also not sure how to test for the changes in `downsample.py`.

There are many existing tests for downsample, so you might consider adapting those.

>     * Should I also test for other commands such as `gather`, `search`?

You should cover all of the code that you changed with tests, as well as providing tests to make sure that the original issue was resolved.

**Comment by keyabarve** [comment]
> > In this issue, I added the function in `utils.py` and modified `compute.py`, `downsample.py` and `watch.py`.
> > ```
> > * Do I need to change any other files?
> > ```
> 
> I don't know; do your current changes address the issue?
> 
I don't think so.
> > ```
> > * When I write tests to check the functionality, which commands should I be testing for?
> > ```
> 
> You should cover all of the code that you changed with tests, as well as providing tests to make sure that the original issue was resolved.
> 
Okay.
> > ```
> > * Since I changed `sourmash compute` to `sourmash sketch` in almost all the places in the other issue I was working on, do I still need to test the `--num bounds` for `compute`?
> > ```
> 
> That's a good question. Please note it in the top of the PR and it can be part of the review.
> 
Sure.
> > ```
> > * I’ll surely write some tests for the `watch` command. I’m also not sure how to test for the changes in `downsample.py`.
> > ```
> 
> There are many existing tests for downsample, so you might consider adapting those.
> 
Alright.
> > ```
> > * Should I also test for other commands such as `gather`, `search`?
> > ```
> 
> You should cover all of the code that you changed with tests, as well as providing tests to make sure that the original issue was resolved.
>
Oh okay.

**Comment by keyabarve** [comment]
I have a few questions/comments:
- I have changed `args.num` to `args.num_hashes` in the `downsample` function of `src/sourmash/sig/__main__.py`. 
- I also wrote tests in `tests/test_sourmash.py` for the `watch` command. 
- Since I have changed `compute` to `sketch` in most of the places in `tests/test_sourmash.py`, I was wondering whether or not I should write the tests for the `compute` command. 
- If so, then should I write it in `tests/test_sourmash_compute.py`? 
- I also wanted to know if it would be okay to write the tests for the `downsample` command in `tests/test_cmd_signature.py`?

File: src/sourmash/cli/utils.py [hunk 1]
```diff
@@ -116,4 +116,27 @@ def add_scaled_arg(parser, default=None):
...
+
+def check_num_bounds(arg):
+    actual_min_val = 0
+    min_val = 50
+    max_val = 50000
+
+    f = int(arg)
+
+    if f < actual_min_val:
+        raise argparse.ArgumentTypeError(f"ERROR: --num value must be positive")
```
Comment (ctb) [inline src/sourmash/cli/utils.py]: this should be `--num-hashes`, right?

**Comment by ctb** [comment]
> I have a few questions/comments:
> 
>     * Since I have changed `compute` to `sketch` in most of the places in `tests/test_sourmash.py`, I was wondering whether or not I should write the tests for the `compute` command.

I think it would be good to write one test, to make sure the new code is actually being used; but since `compute` is being removed in 5.0, you don't need to test it as thoroughly as code that is remaining.

>     * I also wanted to know if it would be okay to write the tests for the `downsample` command in `tests/test_cmd_signature.py`?

Please write the tests wherever you think best, and if you think they could go in multiple places, indicate that in the PR description as a question you have, together with the list of potential locations and why you think they might fit there better than where you put them. Then the code reviewer can address it in their review.

File: src/sourmash/cli/utils.py (hunk 1 above)
Comment (keyabarve) [reply src/sourmash/cli/utils.py]: Oh right, I'll change that.

**Comment by keyabarve** [comment]
> > I have a few questions/comments:
> > ```
> > * Since I have changed `compute` to `sketch` in most of the places in `tests/test_sourmash.py`, I was wondering whether or not I should write the tests for the `compute` command.
> > ```
> 
> I think it would be good to write one test, to make sure the new code is actually being used; but since `compute` is being removed in 5.0, you don't need to test it as thoroughly as code that is remaining.
> 
Alright!
> > ```
> > * I also wanted to know if it would be okay to write the tests for the `downsample` command in `tests/test_cmd_signature.py`?
> > ```
> 
> Please write the tests wherever you think best, and if you think they could go in multiple places, indicate that in the PR description as a question you have, together with the list of potential locations and why you think they might fit there better than where you put them. Then the code reviewer can address it in their review.
>
Okay!

**Comment by keyabarve** [comment]
@ctb This is ready for review.

File: src/sourmash/cli/compute.py [hunk 2]
```diff
@@ -53,10 +53,10 @@ def subparser(subparsers):
         type=ksize_parser,
         help='comma-separated list of k-mer sizes; default=%(default)s'
     )
-    sketch_args.add_argument(
-        '-n', '--num-hashes', type=int, default=500,
-        help='number of hashes to use in each sketch; default=%(default)i'
-    )
+    # sketch_args.add_argument(
```
Comment (ctb) [inline src/sourmash/cli/compute.py]: please remove

**Review by ctb — CHANGES_REQUESTED** [review]
I did a preliminary review and noticed some leftover comments that should be removed.

(I would like to revisit this later to do a larger review, might take me a bit of time.)

File: src/sourmash/cli/compute.py (hunk 2 above)
Comment (keyabarve) [reply src/sourmash/cli/compute.py]: Alright!

**Comment by keyabarve** [comment]
@ctb Please review.

File: src/sourmash/cli/sig/downsample.py [hunk 3]
```diff
@@ -1,7 +1,7 @@
 """downsample one or more signatures"""
 
-from sourmash.cli.utils import (add_moltype_args, add_ksize_arg,
-                                add_picklist_args)
+from sourmash.cli.utils import (add_moltype_args, add_ksize_arg, add_num_arg,
```
Comment (ctb) [inline src/sourmash/cli/sig/downsample.py]: duplicate import of `add_num_arg` here - pls remove

**Review by ctb — CHANGES_REQUESTED** [review]
One minor thing to be cleaned up, otherwise looks good!

**Comment by ctb** [comment]
(oh, and you should of course update from latest too :)

**Comment by keyabarve** [comment]
> (oh, and you should of course update from latest too :)
>
Yes, I'll do that!

**Comment by keyabarve** [comment]
@ctb Please review.

**Review by ctb — APPROVED** [review]

**Comment by ctb** [comment]
thank you!

**Pull request merged**